modules/gapi/src/streaming/onevpl/demux/async_mfp_demux_data_provider.cpp

```diff
@@ -0,0 +1,783 @@
+                                    ", DataLength: "  << staging_stream->DataLength);
+
+            worker_buffer_count = produce_worker_data(staging_stream->Data,
+                                                      createCOMPtrGuard(buffer_ptr),
+                                                      std::move(staging_stream));
+        }
+    } else {
+        GAPI_LOG_WARNING(nullptr, "[" << this << "] callback failed"
+                                  ", status: " << std::to_string(HRESULT_CODE(status)) <<
+                                  ", stream flags: " << stream_flag <<
+                                  ", sample: " << sample_ptr);
+    }
+
+    hr = request_next(hr, stream_flag, worker_buffer_count);
+    return hr;
+}
+
+size_t MFPAsyncDemuxDataProvider::get_locked_buffer_size() const {
+    std::unique_lock<std::mutex> l(buffer_storage_mutex);
+    return worker_locked_buffer_storage.size();
+}
+
+STDMETHODIMP MFPAsyncDemuxDataProvider::OnEvent(DWORD, IMFMediaEvent *) {
+    return S_OK;
+}
+
+STDMETHODIMP MFPAsyncDemuxDataProvider::OnFlush(DWORD) {
+    provider_state.store(State::Exhausted);
+    buffer_storage_non_empty_cond.notify_all();
+    return S_OK;
+}
+
+void MFPAsyncDemuxDataProvider::flush() {
+    if(source_reader) {
+        GAPI_LOG_INFO(nullptr, "[" << this << "] set flush");
+        source_reader->Flush(static_cast<DWORD>(MF_SOURCE_READER_ALL_STREAMS));
+    }
+
+    size_t iterations = 1;
+    while (provider_state.load() != State::Exhausted) {
+        GAPI_LOG_DEBUG(nullptr, "[" << this << "] is waiting for flush finishing, "
+                                "iteration: " <<iterations);
+        iterations++;
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+
+    GAPI_LOG_INFO(nullptr, "[" << this << "] has flushed in iteration: " <<
+                           iterations);
+}
+
+HRESULT MFPAsyncDemuxDataProvider::request_next(HRESULT hr,
+                                                DWORD stream_flag,
+                                                size_t worker_buffer_count) {
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] status: " <<
+                            std::to_string(HRESULT_CODE(hr)) <<
+                            ", stream flags: " << stream_flag <<
+                            ", worker buffer count: (" << worker_buffer_count <<
+                            "/" << keep_preprocessed_buf_count << ")");
+    // check gap in stream
+    if (stream_flag & MF_SOURCE_READERF_STREAMTICK ) {
+        GAPI_LOG_INFO(nullptr, "[" << this << "] stream gap detected");
+        return hr;
+    }
+
+    if (FAILED(hr)) {
+        GAPI_LOG_WARNING(nullptr, "[" << this << "] callback error "
+                                  ", status: " << std::to_string(HRESULT_CODE(hr)) <<
+                                  ", stream flags: " << stream_flag);
+    }
+
+    // put on worker buffers available ready
+    if (worker_buffer_count < keep_preprocessed_buf_count) {
```

Comment:
Why don't you get sizes here directly? If they are changed asynchronously (if it's possible) will it affect anything?

Reply:
>> Why don't you get sizes here directly?

to avoid redundant lock for mutex

>>If they are changed asynchronously (if it's possible)

actual size can be changed (it application grabs frames through `fetch_bitstream`) and it is not supposed to be strong condition here - just keep several preprocessed buffers for optimization purposes.  Because actual drain of buffers (after check its size under lock) may happen AFTER this condition false and we would not process with request next frame here.

The main idea is: move all copy & other heavy work under `OnReadSamle` callback and make `fetch_bitstream` as thin as possible.
Callback `onReadSample` is receiving demultiplexed video sample, putting it to samples buffer (to release it later), and copying samples buffer data to MFX bitstream data... and if MFX bitstrream data portions count is less then `keep_preprocessed_buf_count` constant and no EOF then request next portion again by using `request_next`.
Simulataneously application level by using `fetch_bistream`  with atomic buffer exchange is grabbing  MFX bitstream data array which is supposed to be filling in `onReadSample`. sometimes `onReadSample` stops and reject to `request_next` according to this condition, then `fetch_bitstream` will recharge it by calling `ReadSample` and will wait for complete at least one samples is ready and chain for filling `keep_preprocessed_buf_count` will be released

It is possible to introduce some `supporter`-thread or any other constant `keep_preprocessed_buf_count` availability checker but i think it will require more strong synchronization guarantees and will make `fetch_bitstream` more heavy weight

Reply:
Got it, thanks for explanation!